hub/subscribe.go

```diff
@@ -17,41 +17,60 @@ func (h *Hub) SubscribeHandler(w http.ResponseWriter, r *http.Request) {
 		panic("The Response Writer must be an instance of Flusher.")
 	}
 
-	subscriber, updateChan, ok := h.initSubscription(w, r)
+	subscriber, pipe, ok := h.initSubscription(w, r)
 	if !ok {
 		return
 	}
 	defer h.cleanup(subscriber)
 
-	for {
-		if h.options.HeartbeatInterval == time.Duration(0) {
+	if h.options.HeartbeatInterval == time.Duration(0) {
+		for {
 			// No heartbeat defined, just block
-			serializedUpdate, open := <-updateChan
-			if !open {
-				return
+			update, err := pipe.Read()
+			if err != nil {
+				if err == ErrClosedPipe {
+					return
+				}
+				log.Error(err)
+			} else {
+				h.publish(newSerializedUpdate(update), subscriber, w, r)
 			}
-			h.publish(serializedUpdate, subscriber, w, r)
-
-			continue
 		}
-
-		select {
-		case serializedUpdate, open := <-updateChan:
-			if !open {
-				return
+	} else {
+		updates := make(chan *Update)
+		go func() {
```

Comment:
My construct was designed to not spawn a useless goroutine (goroutines can become costly on high traffic services), could you switch back to the previous construct?
Also, it makes the code simpler.

Reply:
Because of the `select` (to create a race between updates and heartbeat), I have to create an intermediate chan.

For the record, the object Pipe is just a Chan that gracefully handle Closing (reader and writer can close the chan) while a standard chan will panic if someone try to write on it.

I've 3 options:
- keep this goroutine
- add a method `ReadChan` in Pipe that return the chan
- remove the Pipe thing, `Stream::CreatePipe` returns a chan, and add a method `Stream::CloseChan` to gracefully close the chan from reader and writers